From 0a77d72c1b4c8085d432eccc8bcf38ec5d0f4752 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Wed, 27 Jul 2016 21:27:59 +0200
Subject: [PATCH 023/217] printk: irq_pipeline: safe printk with IRQ pipelining

Defer printk output not to wreck the hard interrupt state, cause
unacceptable latency, or risk a deadlock, if either printk() is called
from the oob stage or IRQs are hard disabled on entry (which covers
the case of running over the root stage holding a hard lock).

Deferral is obtained by following the printk-from-NMI path when any of
the above conditions is true. This relies on the ability for the oob
stage to submit work requests to the in-band stage using the
irq_work_queue() interface.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/linux/printk.h            |  2 +-
 kernel/printk/internal.h          | 19 +++++++++++++++++++
 kernel/printk/printk.c            | 30 +++++++++++++++++++++---------
 kernel/printk/printk_ringbuffer.c | 10 +++++-----
 kernel/printk/printk_ringbuffer.h | 12 ++++++++++++
 kernel/printk/printk_safe.c       |  3 ++-
 6 files changed, 60 insertions(+), 16 deletions(-)

diff --git a/include/linux/printk.h b/include/linux/printk.h
index 712307033..96c5c03b9 100644
--- a/include/linux/printk.h
+++ b/include/linux/printk.h
@@ -709,7 +709,7 @@ struct pi_entry {
 				      DEFAULT_RATELIMIT_INTERVAL,	\
 				      DEFAULT_RATELIMIT_BURST);		\
 									\
-	if (__ratelimit(&_rs))						\
+	if (running_oob() || __ratelimit(&_rs))				\
 		printk(fmt, ##__VA_ARGS__);				\
 })
 #else
diff --git a/kernel/printk/internal.h b/kernel/printk/internal.h
index 5eef70000..46fa47935 100644
--- a/kernel/printk/internal.h
+++ b/kernel/printk/internal.h
@@ -169,6 +169,25 @@ static inline void nbcon_kthread_wake(struct console *con)
 	rcuwait_wake_up(&con->rcuwait); /* LMM(nbcon_kthread_wake:A) */
 }
 
+#ifdef CONFIG_IRQ_PIPELINE
+
+extern bool irq_pipeline_active;
+
+static inline bool printk_stage_safe(void)
+{
+	return running_inband() &&
+		(!hard_irqs_disabled() || !irq_pipeline_active);
+}
+
+#else
+
+static inline bool printk_stage_safe(void)
+{
+	return true;
+}
+
+#endif	/* CONFIG_IRQ_PIPELINE */
+
 #else
 
 #define PRINTK_PREFIX_MAX	0
diff --git a/kernel/printk/printk.c b/kernel/printk/printk.c
index ab3473b31..6d1555caf 100644
--- a/kernel/printk/printk.c
+++ b/kernel/printk/printk.c
@@ -2101,10 +2101,10 @@ static u8 *__printk_recursion_counter(void)
 	bool success = true;				\
 							\
 	typecheck(u8 *, recursion_ptr);			\
-	local_irq_save(flags);				\
+	prb_irq_save(flags);				\
 	(recursion_ptr) = __printk_recursion_counter();	\
 	if (*(recursion_ptr) > PRINTK_MAX_RECURSION) {	\
-		local_irq_restore(flags);		\
+		prb_irq_restore(flags);			\
 		success = false;			\
 	} else {					\
 		(*(recursion_ptr))++;			\
@@ -2117,7 +2117,7 @@ static u8 *__printk_recursion_counter(void)
 	do {						\
 		typecheck(u8 *, recursion_ptr);		\
 		(*(recursion_ptr))--;			\
-		local_irq_restore(flags);		\
+		prb_irq_restore(flags);			\
 	} while (0)
 
 int printk_delay_msec __read_mostly;
@@ -2237,9 +2237,6 @@ int vprintk_store(int facility, int level,
 	int ret = 0;
 	u64 ts_nsec;
 
-	if (!printk_enter_irqsave(recursion_ptr, irqflags))
-		return 0;
-
 	/*
 	 * Since the duration of printk() can vary depending on the message
 	 * and state of the ringbuffer, grab the timestamp now so that it is
@@ -2248,8 +2245,6 @@ int vprintk_store(int facility, int level,
 	 */
 	ts_nsec = local_clock();
 
-	caller_id = printk_caller_id();
-
 	/*
 	 * The sprintf needs to come first since the syslog prefix might be
 	 * passed in as a parameter. An extra byte must be reserved so that
@@ -2273,6 +2268,12 @@ int vprintk_store(int facility, int level,
 	if (dev_info)
 		flags |= LOG_NEWLINE;
 
+	/* Disable interrupts as late as possible. */
+	if (!printk_enter_irqsave(recursion_ptr, irqflags))
+		return 0;
+
+	caller_id = printk_caller_id();
+
 	if (flags & LOG_CONT) {
 		prb_rec_init_wr(&r, reserve_size);
 		if (prb_reserve_in_last(&e, prb, &r, caller_id, PRINTKRB_RECORD_MAX)) {
@@ -2369,6 +2370,12 @@ asmlinkage int vprintk_emit(int facility, int level,
 	if (other_cpu_in_panic() && !panic_triggering_all_cpu_backtrace)
 		return 0;
 
+	if (unlikely(!printk_stage_safe())) {
+		printed_len = vprintk_store(facility, level, dev_info, fmt, args);
+		defer_console_output();
+		return printed_len;
+	}
+
 	printk_get_console_flush_type(&ft);
 
 	/* If called from the scheduler, we can not call up(). */
@@ -2506,7 +2513,12 @@ void raw_vprintk(const char *fmt, va_list ap)
 	char buf[256];
 	size_t n;
 
-	if (raw_console == NULL || console_suspended)
+	/*
+	 * Revisit: we should rely on the write_atomic() interface to
+	 * remove the need for write_raw().
+	 */
+	if (raw_console == NULL ||
+	    !console_is_usable(raw_console, raw_console->flags, false))
 		return;
 
         touch_nmi_watchdog();
diff --git a/kernel/printk/printk_ringbuffer.c b/kernel/printk/printk_ringbuffer.c
index 88e8f3a61..6084b5f25 100644
--- a/kernel/printk/printk_ringbuffer.c
+++ b/kernel/printk/printk_ringbuffer.c
@@ -1372,12 +1372,12 @@ bool prb_reserve_in_last(struct prb_reserved_entry *e, struct printk_ringbuffer
 	struct prb_desc *d;
 	unsigned long id;
 
-	local_irq_save(e->irqflags);
+	prb_irq_save(e->irqflags);
 
 	/* Transition the newest descriptor back to the reserved state. */
 	d = desc_reopen_last(desc_ring, caller_id, &id);
 	if (!d) {
-		local_irq_restore(e->irqflags);
+		prb_irq_restore(e->irqflags);
 		goto fail_reopen;
 	}
 
@@ -1614,12 +1614,12 @@ bool prb_reserve(struct prb_reserved_entry *e, struct printk_ringbuffer *rb,
 	 * interrupts during the reserve/commit window in order to minimize
 	 * the likelihood of this happening.
 	 */
-	local_irq_save(e->irqflags);
+	prb_irq_save(e->irqflags);
 
 	if (!desc_reserve(rb, &id)) {
 		/* Descriptor reservation failures are tracked. */
 		atomic_long_inc(&rb->fail);
-		local_irq_restore(e->irqflags);
+		prb_irq_restore(e->irqflags);
 		goto fail;
 	}
 
@@ -1724,7 +1724,7 @@ static void _prb_commit(struct prb_reserved_entry *e, unsigned long state_val)
 	}
 
 	/* Restore interrupts, the reserve/commit window is finished. */
-	local_irq_restore(e->irqflags);
+	prb_irq_restore(e->irqflags);
 }
 
 /**
diff --git a/kernel/printk/printk_ringbuffer.h b/kernel/printk/printk_ringbuffer.h
index 4ef81349d..8ad684ebc 100644
--- a/kernel/printk/printk_ringbuffer.h
+++ b/kernel/printk/printk_ringbuffer.h
@@ -239,6 +239,18 @@ enum desc_state {
 #define DESC0_ID(ct_bits)	DESC_ID(-(_DESCS_COUNT(ct_bits) + 1))
 #define DESC0_SV(ct_bits)	DESC_SV(DESC0_ID(ct_bits), desc_reusable)
 
+/*
+ * When interrupt pipelining is enabled, we want the critical sections
+ * to be protected against preemption by out-of-band code.
+ */
+#ifdef CONFIG_IRQ_PIPELINE
+#define prb_irq_save(__flags)		do { (__flags) = hard_local_irq_save(); } while (0)
+#define prb_irq_restore(__flags)	hard_local_irq_restore(__flags)
+#else
+#define prb_irq_save(__flags)		local_irq_save(__flags)
+#define prb_irq_restore(__flags)	local_irq_restore(__flags)
+#endif
+
 /*
  * Define a ringbuffer with an external text data buffer. The same as
  * DEFINE_PRINTKRB() but requires specifying an external buffer for the
diff --git a/kernel/printk/printk_safe.c b/kernel/printk/printk_safe.c
index e6198da7c..374368ca3 100644
--- a/kernel/printk/printk_safe.c
+++ b/kernel/printk/printk_safe.c
@@ -9,6 +9,7 @@
 #include <linux/cpumask.h>
 #include <linux/printk.h>
 #include <linux/kprobes.h>
+#include <linux/irqstage.h>
 
 #include "internal.h"
 
@@ -66,7 +67,7 @@ asmlinkage int vprintk(const char *fmt, va_list args)
 	 * Use the main logbuf even in NMI. But avoid calling console
 	 * drivers that might have their own locks.
 	 */
-	if (is_printk_legacy_deferred())
+	if (!printk_stage_safe() || is_printk_legacy_deferred())
 		return vprintk_deferred(fmt, args);
 
 	/* No obstacles. */
-- 
2.30.2

